File: library/src/main/java/me/proxer/library/api/user/UserInfoEndpoint.java
package me.proxer.library.api.user;

import lombok.experimental.Accessors;
import me.proxer.library.api.Endpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entity.user.UserInfo;

import javax.annotation.Nullable;

/**
 * Endpoint for requesting various information of an user.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class UserInfoEndpoint implements Endpoint<UserInfo> {

    private final InternalApi internalApi;

    @Nullable
    private final String userId;

    @Nullable
    private final String username;

    UserInfoEndpoint(final InternalApi internalApi, @Nullable final String userId,
                     @Nullable final String username) {
        if (userId == null && username == null) {
            throw new IllegalArgumentException("You must pass either an userId or an username.");
        }

        this.internalApi = internalApi;
        this.userId = userId;
        this.username = username;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ProxerCall<UserInfo> build() {
        return internalApi.userInfo(userId, username);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/user/UserInfoEndpoint.kt
package me.proxer.library.api.user

import me.proxer.library.ProxerCall
import me.proxer.library.api.Endpoint
import me.proxer.library.entity.user.UserInfo

/**
 * Endpoint for requesting various information of an user.
 *
 * @author Ruben Gees
 */
class UserInfoEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val userId: String?,
    private val username: String?
) : Endpoint<UserInfo> {

    init {
        if (userId.isNullOrBlank() && username.isNullOrBlank()) {
            throw IllegalArgumentException("You must pass either an userId or an username.")
        }
    }

    override fun build(): ProxerCall<UserInfo> {
        return internalApi.userInfo(userId, username)
    }
}
